ably/realtime_channel.go

```diff
@@ -240,19 +246,46 @@ func (c *RealtimeChannel) lockAttach(err error) (result, error) {
 // returns with an error, but the operation carries on in the background and
 // the channel may eventually be detached anyway.
 func (c *RealtimeChannel) Detach(ctx context.Context) error {
+	prevChannelState := c.State()
 	res, err := c.detach()
 	if err != nil {
 		return err
 	}
-	return res.Wait(ctx)
+	resultErr := make(chan error, 1)
+	go func() {
+		resultErr <- res.Wait(ctx)
+	}()
+
+	detachTimeout := c.client.Connection.opts.realtimeRequestTimeout()
+	select {
+	case err := <-resultErr: // RTL5e
+		return err
+	case <-time.After(detachTimeout): // RTL5f
+		err := newError(ErrTimeoutError, errors.New("timed out before detaching channel"))
+		c.setState(prevChannelState, err)
+		return err
+	}
```

Comment:
Now that I think of it, we shouldn't use `context.WithTimeout` directly, because it bypasses ClientOptions.After and thus prevents us from controlling timers from tests.

We need a custom one, something like this:

```go
func (o *clientOptions) withTimeout(ctx context.Context, timeout time.Duration) (context.Context, context.CancelFunc) {
	ctx, cancel := context.WithCancel(ctx)
	go func() {
		<-o.After(ctx, timeout)
		cancel()
	}()
	return ctx, cancel
}
```

Reply:
Okay, since the second solution already contains goroutine, I can directly use that function in the current implementation.
Replace `time.After(detachTimeout)` with `c.opts().After(context.Background(), detachTimeout)`

Reply:
It seems, once I change it, current transitioner tests are breaking, also as far as detach is concerned, it seems to be working without the need of injecting an extra `option.After` function from outside.

Reply:
So, I will keep it the way it is for the time being

Reply:
> It seems, once I change it, current transitioner tests are breaking, also as far as detach is concerned, it seems to be working without the need of injecting an extra option.After function from outside.

As discussed, I've addressed this in f769f68779b7c85e35164df24b67bcc94ed82232 and c6d3c3d3d786b866b8d520e480fb0dd5a691e372.

Reply:
I think we can close this one, since it's already addressed